File: bindings/rust/src/safe_graph.rs
// SPDX-License-Identifier: PMPL-1.0-or-later
// Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <jonathan.jewell@open.ac.uk>

//! Safe directed graph via libproven FFI.
//!
//! Provides a directed graph with edge operations.
//! All operations delegate to Idris 2 verified code.

use crate::core::{self, Error, Result};
use crate::ffi;

/// A directed graph with a fixed number of nodes.
///
/// The underlying memory is managed by libproven. The graph is freed
/// when this struct is dropped.
pub struct Graph {
    ptr: *mut ffi::Graph,
}

// SAFETY: The underlying graph is managed by libproven.
unsafe impl Send for Graph {}

impl Graph {
    /// Create a new directed graph with the given number of nodes.
    ///
    /// Nodes are indexed from 0 to `node_count - 1`.
    pub fn new(node_count: usize) -> Result<Self> {
        // SAFETY: proven_graph_create takes a value-type usize;
        // always safe to call.
        let ptr = unsafe { ffi::proven_graph_create(node_count) };
        if ptr.is_null() {
            return Err(Error::AllocationFailed);
        }
        Ok(Graph { ptr })
    }

    /// Add a directed edge from node `from` to node `to`.
    ///
    /// Returns `Err(Error::OutOfBounds)` if either node index is invalid.
    pub fn add_edge(&mut self, from: usize, to: usize) -> Result<()> {
        // SAFETY: self.ptr is valid (checked at construction).
        let status = unsafe {
            ffi::proven_graph_add_edge(self.ptr, from, to)
        };
        core::status_to_result(status)
    }

    /// Check if a directed edge exists from node `from` to node `to`.
    pub fn has_edge(&self, from: usize, to: usize) -> bool {
        // SAFETY: self.ptr is valid (checked at construction).
        unsafe { ffi::proven_graph_has_edge(self.ptr, from, to) }
    }
}

impl Drop for Graph {
    fn drop(&mut self) {
        // SAFETY: self.ptr was allocated by proven_graph_create
        // and has not been freed yet.
        unsafe {
            ffi::proven_graph_free(self.ptr);
        }
    }
}

File: bindings/zig/src/safe_graph.zig
// SPDX-License-Identifier: PMPL-1.0-or-later
// Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <jonathan.jewell@open.ac.uk>
//
// Proven SafeGraph - FFI bindings to libproven graph operations.
// All computation is performed in verified Idris 2 code via libproven.

const std = @import("std");
const c = @cImport(@cInclude("proven.h"));

/// Error types for graph operations.
pub const GraphError = error{
    CreationFailed,
    OutOfBounds,
    ProvenError,
};

/// Directed graph backed by libproven.
pub const Graph = struct {
    ptr: *c.ProvenGraph,

    /// Add directed edge from -> to via libproven.
    pub fn addEdge(self: Graph, from: usize, to: usize) GraphError!void {
        const status = c.proven_graph_add_edge(self.ptr, from, to);
        return switch (status) {
            c.PROVEN_OK => {},
            c.PROVEN_ERR_OUT_OF_BOUNDS => error.OutOfBounds,
            else => error.ProvenError,
        };
    }

    /// Check if directed edge exists via libproven.
    pub fn hasEdge(self: Graph, from: usize, to: usize) bool {
        return c.proven_graph_has_edge(self.ptr, from, to);
    }

    /// Free graph via libproven.
    pub fn deinit(self: Graph) void {
        c.proven_graph_free(self.ptr);
    }
};

/// Create a directed graph via libproven.
/// node_count: Number of nodes (max 10,000).
pub fn create(node_count: usize) GraphError!Graph {
    const ptr = c.proven_graph_create(node_count);
    if (ptr == null) return error.CreationFailed;
    return Graph{ .ptr = ptr.? };
}

test "create and add edges" {
    const g = try create(5);
    defer g.deinit();
    try g.addEdge(0, 1);
    try g.addEdge(1, 2);
    try std.testing.expect(g.hasEdge(0, 1));
    try std.testing.expect(!g.hasEdge(2, 0));
}
